File: src/java/br/com/lett/crawlernode/crawlers/corecontent/brasil/BrasilRennerCrawler.java
package br.com.lett.crawlernode.crawlers.corecontent.brasil;

import br.com.lett.crawlernode.core.fetcher.FetchMode;
import br.com.lett.crawlernode.core.fetcher.ProxyCollection;
import br.com.lett.crawlernode.core.fetcher.methods.ApacheDataFetcher;
import br.com.lett.crawlernode.core.fetcher.methods.FetcherDataFetcher;
import br.com.lett.crawlernode.core.fetcher.methods.JsoupDataFetcher;
import br.com.lett.crawlernode.core.fetcher.models.Request;
import br.com.lett.crawlernode.core.fetcher.models.Response;
import br.com.lett.crawlernode.core.models.Card;
import br.com.lett.crawlernode.core.models.Parser;
import br.com.lett.crawlernode.core.models.Product;
import br.com.lett.crawlernode.core.models.ProductBuilder;
import br.com.lett.crawlernode.core.session.Session;
import br.com.lett.crawlernode.core.task.impl.Crawler;
import br.com.lett.crawlernode.util.CrawlerUtils;
import br.com.lett.crawlernode.util.JSONUtils;
import br.com.lett.crawlernode.util.Logging;
import com.google.common.collect.Sets;
import exceptions.MalformedPricingException;
import exceptions.OfferException;
import models.Offer;
import models.Offers;
import models.RatingsReviews;
import models.pricing.*;
import org.json.JSONArray;
import org.json.JSONObject;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Set;

;


/**
 * Date: 14/07/20
 *
 * @author Fellype Layunne
 */
public class BrasilRennerCrawler extends Crawler {

   private static String SELLER_NAME = "Renner";

   public Set<String> cards = Sets.newHashSet(Card.VISA.toString(), Card.MASTERCARD.toString(),
      Card.ELO.toString(), Card.DINERS.toString(), Card.HIPER.toString(), Card.AMEX.toString());

   public BrasilRennerCrawler(Session session) {
      super(session);
      super.config.setParser(Parser.HTML);
      super.config.setFetcher(FetchMode.JSOUP);
   }

   @Override
   protected Response fetchResponse() {
      Request request = Request.RequestBuilder.create()
         .setUrl(session.getOriginalURL())
         .setProxyservice(List.of(ProxyCollection.LUMINATI_SERVER_BR, ProxyCollection.BUY))
         .build();

      Response response = CrawlerUtils.retryRequestWithListDataFetcher(request, List.of(new ApacheDataFetcher(), new FetcherDataFetcher(), new JsoupDataFetcher()), session, "get");

      return response;
   }

   @Override
   public List<Product> extractInformation(Document document) throws Exception {
      List<Product> products = new ArrayList<>();

      JSONObject pageJson = CrawlerUtils.selectJsonFromHtml(document, "#__NEXT_DATA__", null, null, false, false);
      JSONObject productJson = JSONUtils.getValueRecursive(pageJson, "props.pageProps.product", JSONObject.class);

      if (productJson != null && !productJson.isEmpty()) {
         String internalId = productJson.optString("skuId");
         String internalPid = productJson.optString("productId");
         String description = getDescription(pageJson);
         String displayName = productJson.optString("displayName");
         List<String> categories = CrawlerUtils.crawlCategories(document, "ul[aria-label='breadcrumb'] li:not(:first-child):not(:last-child) a", true);
         boolean available = productJson.optBoolean("purchasable");
         RatingsReviews ratingsReviews = ratingsReviews(productJson);
         JSONObject productResponse = fetchProductResponse(internalId, internalPid);
         String primaryImage = getPrimaryImage(productResponse);
         List<String> secondaryImages = getSecondaryImages(productResponse);
         Offers offers = available ? scrapOffers(productResponse, internalId, internalPid) : new Offers();

         Elements variations = document.select(".ProductAttributes_ProductAttributeWrapper__aRm_h");

         if (variations.size() > 0) {
            for (Element e : variations) {
               internalId = CrawlerUtils.scrapStringSimpleInfoByAttribute(e, ".ProductAttributes_inputCheckbox__5Y371", "data-sku");
               String variationName = CrawlerUtils.scrapStringSimpleInfoByAttribute(e, ".ProductAttributes_inputCheckbox__5Y371", "data-name");
               String productName = displayName + " " + variationName;

               Product product = ProductBuilder.create()
                  .setUrl(session.getOriginalURL())
                  .setInternalId(internalId)
                  .setInternalPid(internalPid)
                  .setName(productName)
                  .setPrimaryImage(primaryImage)
                  .setSecondaryImages(secondaryImages)
                  .setDescription(description)
                  .setCategories(categories)
                  .setRatingReviews(ratingsReviews)
                  .setOffers(offers)
                  .build();
               products.add(product);
            }
         } else {
            Product product = ProductBuilder.create()
               .setUrl(session.getOriginalURL())
               .setInternalId(internalId)
               .setInternalPid(internalPid)
               .setName(displayName)
               .setPrimaryImage(primaryImage)
               .setSecondaryImages(secondaryImages)
               .setDescription(description)
               .setCategories(categories)
               .setRatingReviews(ratingsReviews)
               .setOffers(offers)
               .build();
            products.add(product);
         }
      } else {
         Logging.printLogDebug(logger, session, "Not a product page " + this.session.getOriginalURL());
      }
      return products;
   }

   private String getPrimaryImage(JSONObject productResponse) {
      String primaryImage = null;
      JSONArray mediaSets = JSONUtils.getValueRecursive(productResponse, "mediaSets", JSONArray.class);
      if (mediaSets != null && !mediaSets.isEmpty()) {
         JSONObject images = (JSONObject) mediaSets.opt(0);
         if (images != null && !images.isEmpty()) {
            primaryImage = CrawlerUtils.completeUrl(images.optString("largeImageUrl"), "https", "");
         }
      }
      return primaryImage;
   }

   private List<String> getSecondaryImages(JSONObject object) {
      JSONArray mediaSets = JSONUtils.getValueRecursive(object, "mediaSets", JSONArray.class);
      JSONObject images;
      List<String> secondaryImages = new ArrayList<>();
      if (mediaSets != null && !mediaSets.isEmpty()) {
         for (int i = 0; i < mediaSets.length(); i++) {
            images = (JSONObject) mediaSets.opt(i);
            secondaryImages.add(CrawlerUtils.completeUrl(images.optString("largeImageUrl"), "https", ""));
         }
      }
      if (secondaryImages.size() > 0) {
         secondaryImages.remove(0);
      }
      return secondaryImages;
   }

   private String getDescription(JSONObject object) {
      String description = null;
      if (object != null && !object.isEmpty()) {
         JSONObject descriptionObj = JSONUtils.getValueRecursive(object, "props.pageProps.content", JSONObject.class);
         if (descriptionObj != null && !descriptionObj.isEmpty()) {
            JSONArray mainContent = JSONUtils.getValueRecursive(descriptionObj, "mainContent", JSONArray.class);
            if (mainContent != null && !mainContent.isEmpty()) {
               JSONObject metaDescription = JSONUtils.getValueRecursive(mainContent, "0.record.attributes", JSONObject.class);
               if (metaDescription != null && !metaDescription.isEmpty()) {
                  description = metaDescription.optString("prop.product.metaDescription");
               }
            }
         }
      }
      return description;
   }

   private JSONObject fetchProductResponse(String internalId, String internalPid) {
      String urlVariation = "https://www.lojasrenner.com.br/rest/model/lrsa/api/CatalogActor/refreshProductPage?skuId=" + internalId + "&productId=" + internalPid;
      Request request = Request.RequestBuilder.create()
         .setUrl(urlVariation)
         .setProxyservice(List.of(
            ProxyCollection.LUMINATI_SERVER_BR,
            ProxyCollection.NETNUT_RESIDENTIAL_BR,
            ProxyCollection.NETNUT_RESIDENTIAL_BR_HAPROXY))
         .build();
      Response response = CrawlerUtils.retryRequest(request, session, new FetcherDataFetcher(), true);
      return CrawlerUtils.stringToJson(response.getBody());
   }

   private Offers scrapOffers(JSONObject object, String internalId, String internalPid) throws MalformedPricingException, OfferException {
      Offers offers = new Offers();
      if (internalId != null && internalPid != null) {
         Pricing pricing = scrapPricing(object);
         List<String> sales = Collections.singletonList(CrawlerUtils.calculateSales(pricing));

         offers.add(new Offer.OfferBuilder()
            .setUseSlugNameAsInternalSellerId(true)
            .setSellerFullName(this.SELLER_NAME)
            .setMainPagePosition(1)
            .setIsBuybox(false)
            .setIsMainRetailer(true)
            .setPricing(pricing)
            .setSales(sales)
            .build());
      }
      return offers;
   }

   private Pricing scrapPricing(JSONObject object) throws MalformedPricingException {
      Double spotlightPrice = object.optDouble("salePrice");
      Double priceFrom = object.optDouble("listPrice");
      if (spotlightPrice.isNaN()) {
         spotlightPrice = priceFrom;
         priceFrom = null;
      }

      BankSlip bankSlip = CrawlerUtils.setBankSlipOffers(spotlightPrice, null);
      CreditCards creditCards = scrapCreditCards(spotlightPrice);

      return Pricing.PricingBuilder.create()
         .setSpotlightPrice(spotlightPrice)
         .setPriceFrom(priceFrom)
         .setCreditCards(creditCards)
         .setBankSlip(bankSlip)
         .build();

   }

   private CreditCards scrapCreditCards(Double price) throws MalformedPricingException {
      CreditCards creditCards = new CreditCards();
      Installments installments = new Installments();

      installments.add(Installment.InstallmentBuilder.create()
         .setInstallmentNumber(1)
         .setInstallmentPrice(price)
         .setFinalPrice(price)
         .build());

      for (String card : cards) {
         creditCards.add(CreditCard.CreditCardBuilder.create()
            .setBrand(card)
            .setInstallments(installments)
            .setIsShopCard(false)
            .build());
      }

      return creditCards;
   }

   private RatingsReviews ratingsReviews(JSONObject object) {
      RatingsReviews ratingReviews = new RatingsReviews();
      JSONObject review = JSONUtils.getValueRecursive(object, "review", JSONObject.class);
      Integer totalNumOfEvaluations = review.optInt("count");
      Double avgRating = review.optDouble("average");

      ratingReviews.setTotalRating(totalNumOfEvaluations);
      ratingReviews.setAverageOverallRating(avgRating);
      ratingReviews.setTotalWrittenReviews(totalNumOfEvaluations);

      return ratingReviews;
   }
}

File: src/java/br/com/lett/crawlernode/crawlers/corecontent/brasil/BrasilRennerCrawler.kt
package br.com.lett.crawlernode.crawlers.corecontent.brasil

import br.com.lett.crawlernode.core.fetcher.FetchMode
import br.com.lett.crawlernode.core.fetcher.ProxyCollection
import br.com.lett.crawlernode.core.fetcher.models.Request
import br.com.lett.crawlernode.core.models.Card
import br.com.lett.crawlernode.core.models.Product
import br.com.lett.crawlernode.core.models.ProductBuilder
import br.com.lett.crawlernode.core.session.Session
import br.com.lett.crawlernode.core.task.impl.Crawler
import br.com.lett.crawlernode.crawlers.extractionutils.core.TrustvoxRatingCrawler
import br.com.lett.crawlernode.util.*
import exceptions.MalformedPricingException
import models.Offer
import models.Offers
import models.RatingsReviews
import models.pricing.CreditCard.CreditCardBuilder
import models.pricing.CreditCards
import models.pricing.Installment
import models.pricing.Installments
import models.pricing.Pricing
import org.json.JSONObject
import org.jsoup.nodes.Document


/**
 * Date: 14/07/20
 *
 * @author Fellype Layunne
 *
 */
class BrasilRennerCrawler(session: Session) : Crawler(session) {

   init {
      config.fetcher = FetchMode.APACHE
   }

   companion object {
      const val SELLER_NAME: String = "Renner"
   }

   override fun fetch(): Document {
      val request = Request.RequestBuilder.create()
         .setUrl(this.session.originalURL)
         .setProxyservice(
            listOf(
               ProxyCollection.LUMINATI_SERVER_BR
            )
         )
         .build()

      val response = dataFetcher.get(session, request)

      return response.body.toDoc() ?: Document(session.originalURL)
   }

   override fun extractInformation(doc: Document): MutableList<Product> {

      val pageJson = CrawlerUtils.selectJsonFromHtml(doc, "#__NEXT_DATA__", null, null, false, false)
      val productJson = pageJson?.optJSONObject("props")?.optJSONObject("pageProps")?.optJSONObject("product")

      if (isProductPage(productJson)) {
         val products = mutableListOf<Product>()

         val internalId = crawlInternalId(productJson)
         val internalPid = CrawlerUtils.scrapStringSimpleInfoByAttribute(doc, "input[name=product]", "value")
         val urlBase = "https://www.lojasrenner.com.br"
         val urlPath = pageJson?.optJSONObject("props")?.optJSONObject("pageProps")?.optJSONObject("content")?.optJSONArray("mainContent")?.optJSONObject(0)?.optJSONObject("record")?.optJSONObject("attributes")?.optString("prop.product.url")
         val url = "$urlBase$urlPath?sku=$internalId"
         val baseName = productJson?.optString("displayName") ?: ""
         val currentVariantName = productJson?.optString("variants")?.replace("indefinido", "")?.replace("|", " ") ?: ""
         val name = "$baseName $currentVariantName"
         val categories = CrawlerUtils.crawlCategories(doc, "ul[aria-label='breadcrumb'] li:not(:first-child):not(:last-child) a", true)
         val description = pageJson?.optJSONObject("props")?.optJSONObject("pageProps")?.optJSONObject("content")?.optJSONArray("mainContent")?.optJSONObject(0)?.optJSONObject("record")?.optJSONObject("attributes")?.optString("prop.product.metaDescription")
         val images = productJson?.optJSONArray("mediaSets")?.map { "https:${(it as JSONObject).optString("largeImageUrl")}" }
         val offers = if (productJson?.optBoolean("purchasable") == true) scrapOffers(productJson) else Offers()
         val ratings = scrapRating(internalPid, doc)

         val product = ProductBuilder()
            .setUrl(url)
            .setInternalId(internalId)
            .setInternalPid(internalPid)
            .setName(name)
            .setCategories(categories)
            .setPrimaryImage(images?.get(0))
            .setSecondaryImages(images?.let { images.subList(1, it.size) })
            .setDescription(description)
            .setOffers(offers)
            .setRatingReviews(ratings)
            .build()

         products addNonNull product

         return products

      }

      return mutableListOf()
   }

   private fun crawlInternalId(productJson: JSONObject?): String? {
      if (this.session.originalURL.contains("?sku=")) {
         val urlSplit = this.session.originalURL.split("?sku=")
         if (urlSplit.size > 1) {
            val internalId = urlSplit[1]
            if (!internalId.equals("")) {
               return internalId
            }
         }
      }

      return productJson?.optString("skuId")
   }

   private fun scrapInstallments(productJson: JSONObject?): Installments {
      val installments = Installments()

      val installmentNumber = productJson?.optInt("installment")
      val installmentPriceFormatted = productJson?.optString("installmentValueFormatted")
      val installmentPrice = MathUtils.parseDoubleWithComma(installmentPriceFormatted)

      if (installmentNumber != null && installmentPrice != null) {
         val installment = Installment.InstallmentBuilder
            .create()
            .setInstallmentNumber(installmentNumber)
            .setInstallmentPrice(installmentPrice)
            .setFinalPrice((installmentNumber.times(installmentPrice)).toDouble().round())
            .build()

         installments.add(installment)
      }

      return installments
   }

   private fun scrapOffers(productJson: JSONObject): Offers {

      val offers = Offers()

      val priceText = productJson.optString("listPriceFormatted")
      val spotlightText = productJson.optString("salePriceFormatted")

      var priceFrom = MathUtils.parseDoubleWithComma(priceText)
      val spotlightPrice = if (productJson.optDouble("percentDiscount") > 0) {
         MathUtils.parseDoubleWithComma(spotlightText)
      } else {
         priceFrom
      }

      spotlightPrice?.let {
         if (spotlightPrice == priceFrom) {
            priceFrom = null
         }

         val sales = mutableListOf<String>()

         sales addNonNull productJson.optDouble("percentDiscount").toString()

         val otherCars =
            listOf(Card.MASTERCARD, Card.VISA, Card.AMEX, Card.DINERS, Card.ELO,
               Card.HIPERCARD).map { card: Card ->
               try {
                  return@map CreditCardBuilder.create()
                     .setBrand(card.toString())
                     .setIsShopCard(false)
                     .setInstallments(
                        scrapInstallments(productJson)
                     )
                     .build()
               } catch (e: MalformedPricingException) {
                  throw RuntimeException(e)
               }
            }

         val creditCards = CreditCards(otherCars)

         offers.add(
            Offer.OfferBuilder.create()
               .setPricing(
                  Pricing.PricingBuilder.create()
                     .setCreditCards(creditCards)
                     .setSpotlightPrice(spotlightPrice)
                     .setBankSlip(spotlightPrice.toBankSlip())
                     .setPriceFrom(priceFrom)
                     .build()
               )
               .setSales(sales)
               .setIsMainRetailer(true)
               .setIsBuybox(false)
               .setUseSlugNameAsInternalSellerId(true)
               .setSellerFullName(SELLER_NAME)
               .build()
         )
      }

      return offers
   }

   private fun isProductPage(productJson: JSONObject?): Boolean {
      return productJson != null
   }

   private fun scrapRating( internalPid: String?, doc: Document?): RatingsReviews? {
      val trustVox = TrustvoxRatingCrawler(session, "110773", logger)
      return trustVox.extractRatingAndReviews(internalPid, doc, dataFetcher)
   }
}
